codeworld-game-server/src/CodeWorld/GameServer.hs

```diff
@@ -0,0 +1,267 @@
+        modifyMVar_ (games state) (return . HM.delete key)
+  where
+    go g = let players' = filter ((/= mypid) . fst) (players g)
+           in return $ (g { players = players' }, null players')
+
+-- Communication
+
+sendServerMessage :: ServerMessage -> WS.Connection ->  IO ()
+sendServerMessage msg conn = WS.sendTextData conn (T.pack (show msg))
+
+getClientMessage :: WS.Connection -> IO ClientMessage
+getClientMessage conn = do
+    msg <- WS.receiveData conn
+    case readMaybe (T.unpack msg) of
+        Just msg -> return msg
+        Nothing -> fail "Invalid client message"
+
+broadcast :: ServerMessage -> MVar Game -> IO ()
+broadcast msg gameMV = withMVar gameMV $ \game ->
+    forM_ (players game) (sendServerMessage msg . snd)
+
+
+-- Statistics
+
+data CurrentStats = CurrentStats
+    { waitingGames :: !Int
+    , runningGames :: !Int
+    , connections  :: !Int
+    } deriving (Show, Generic)
+instance ToJSON CurrentStats
+
+data TotalStats = TotalStats
+    { totalConnections :: !Int
+    , totalGames       :: !Int
+    } deriving (Show, Generic)
+instance ToJSON TotalStats
+
+tickConnection :: ServerState -> IO ()
+tickConnection state = modifyMVar_ (totalStats state) $ \ts ->
+    return $! ts { totalConnections = totalConnections ts + 1}
+
+tickGame :: ServerState -> IO ()
+tickGame state = modifyMVar_ (totalStats state) $ \ts ->
+    return $! ts { totalGames = totalGames ts + 1}
+
+data ServerStats = ServerStats CurrentStats TotalStats
+-- | merge the fields of 'CurrentStats' and 'TotalStats'
+instance ToJSON ServerStats where
+    toJSON (ServerStats cs ts) = Object (o1 <> o2)
+      where Object o1 = toJSON cs
+            Object o2 = toJSON ts
+
+allGames :: ServerState -> IO [Game]
+allGames state = do
+    gm <- readMVar (games state)
+    mapM readMVar $ HM.elems gm
+
+tally :: [Game]  -> CurrentStats
+tally games = CurrentStats {..}
+  where
+    waitingGames = length [ () | Waiting {} <- games ]
+    runningGames = length [ () | Running {} <- games ]
+    connections  = sum [ length (players g) | g <- games ]
+
+gameStats :: MonadSnap m => ServerState -> m ()
+gameStats state = do
+    cs <- tally <$> liftIO (allGames state)
+    ts <- liftIO $ readMVar (totalStats state)
+    let stats = ServerStats cs ts
+    modifyResponse $ setHeader "Content-Type" "application/json"
+    writeLBS (encode stats)
+
+
+-- Handling logic
+
+-- | Initializes the mutable state of the game server
+initGameServer :: IO ServerState
+initGameServer = do
+    started <- getCurrentTime
+    totalStats <- newMVar (TotalStats 0 0)
+    games <- newMVar HM.empty
+    return $ ServerState {..}
+
+
+-- | A snapp handler
+gameServer :: MonadSnap m => ServerState -> m ()
+gameServer state = runWebSocketsSnap (wsApp state)
+
+wsApp :: ServerState -> WS.ServerApp
+wsApp state pending = do
+    conn <- WS.acceptRequest pending
+    WS.forkPingThread conn 30
+    welcome conn state `catch` \e -> print (e :: SomeException)
```

Comment:
I'm not sure we need anything as complex as you this.  Unless I'm getting something wrong, I think if you just remove this catch, then exceptions will be handled in the normal way, which will write them to Snap's error logs.